packages/rum-core/src/domain/rumEventsCollection/action/trackActions.ts

```diff
@@ -250,3 +270,52 @@ function newPotentialAction(
     },
   }
 }
+
+function addClickToClickChain(state: TrackActionsState, singleClickPotentialAction: PotentialAction): ClickReference {
+  const click: ActionClick = {
+    event: singleClickPotentialAction.base.event,
+    timeStamp: singleClickPotentialAction.base.startClocks.timeStamp,
+    singleClickPotentialAction,
+  }
+  let clickReference = state.currentClickChain && state.currentClickChain.tryAppend(click)
+  // If we failed to add the click to the current click chain, create a new click chain
+  if (!clickReference) {
+    const rageClickPotentialAction = newPotentialAction(state, singleClickPotentialAction.base)
+    const newClickChain = createClickChain<ActionClick>((clicks) => {
+      flushClickChain(clicks, rageClickPotentialAction)
+      stopSubscription.unsubscribe()
+    })
+    const stopSubscription = state.stopObservable.subscribe(newClickChain.stop)
+    // adding a click to a newly created click chain should always succeed
+    clickReference = newClickChain.tryAppend(click)!
+    state.currentClickChain = newClickChain
+  }
+  return clickReference
+}
+
+function flushClickChain(clicks: ActionClick[], rageClickPotentialAction: PotentialAction) {
+  if (isRage(clicks)) {
+    // Merge any click frustration to the rage click action. In practice, it only concerns
+    // 'dead', because any potential 'error' frustration will already be there (as potential action
+    // collect it themselves), and 'rage' won't be on single click potential actions.
+    clicks.forEach((click) => {
+      click.singleClickPotentialAction.frustrations.forEach((frustration) =>
+        rageClickPotentialAction.frustrations.add(frustration)
+      )
+    })
+    // Send the rage click action
+    rageClickPotentialAction.frustrations.add(FrustrationType.RAGE)
+    rageClickPotentialAction.complete(timeStampNow())
+    rageClickPotentialAction.notifyIfComplete()
+  } else {
+    rageClickPotentialAction.discard()
+    // Send an action for each individual click
+    clicks.forEach((click) => click.singleClickPotentialAction.notifyIfComplete())
+  }
+}
+
+const RAGE_CLICK_MIN_COUNT = 3
+function isRage(clicks: ActionClick[]) {
+  // TODO: this condition should be improved to avoid reporting 3-click selection as rage click
+  return clicks.length >= RAGE_CLICK_MIN_COUNT
+}
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏what about moving this check to the click chain? we already have the `3` constant there and other rage click rules there (timing + distance)
We could have something like:
```ts
createPotentialRageClickChain((isRageClickChain, bufferedActions) => {
      ...
})
```